### ISSUE TITLE
precision of convolution

### ISSUE THREAD

**Issue by MOLOjl** [oneapi-src/oneDNN#1221]
hello there，
here is a small test code I write, but the result seems a little wrong. could anyone please tell me why this happen?
```c++
#include <stdio.h>
#include <math.h>

#include "oneapi/dnnl/dnnl.hpp"
#include "oneapi/dnnl/dnnl_debug.h"

#include "example_utils.hpp"

using namespace dnnl;

void convtest(engine::kind engine_kind){
    
    using tag = memory::format_tag;
    using dt = memory::data_type;

    int arg[7] = {3, 227, 227, 11, 4, 96, 0};

    // onednn prepare    
    engine eng(engine_kind, 0);
    stream engine_stream(eng);

    // host prepare
    int batch=100, input_c=arg[0], input_h=arg[1], input_w=arg[2];
    int filter_size=arg[3], stride=arg[4], padding=arg[6];
    int output_c=arg[5];
    int output_h = (input_h - filter_size + 2*padding) / stride + 1; 
    int output_w = (input_w - filter_size + 2*padding) / stride + 1;

    // host data
    std::vector<float> src_v(batch * input_c * input_h * input_w);
    std::vector<float> dst_v(batch * output_c * output_h * output_w);
    std::vector<float> weight_v(output_c * input_c * filter_size * filter_size);
    std::vector<float> bias_v(output_c);

    for (size_t i = 0; i < src_v.size(); ++i)
        src_v[i] = 1.0f;
    for (size_t i = 0; i < weight_v.size(); ++i)
        weight_v[i] = 0.1f;
    for(size_t i = 0; i < bias_v.size(); ++i)
        bias_v[i] = 0;

    
    // tensor dimensions
    memory::dims conv_src_tz = {batch, input_c, input_h, input_h};
    memory::dims conv_weights_tz = {output_c, input_c, filter_size, filter_size};
    memory::dims conv_bias_tz = {output_c};
    memory::dims conv_dst_tz = {batch, output_c, output_h, output_h};
    memory::dims conv_strides = {stride, stride};
    memory::dims conv_padding = {arg[6], arg[6]};

    std::cout<<"input tensor: ";
    for(int i=0; i<3; i++)
        std::cout<<conv_src_tz[i]<<"×";
    std::cout<<conv_src_tz[3]<<"\n";
    
    std::cout<<"filter: ";
    for(int i=0; i<3; i++)
        std::cout<<conv_weights_tz[i]<<"×";
    std::cout<<conv_weights_tz[3]<<"\n";

    // create memory descriptions 
    auto conv_src_md = memory::desc({conv_src_tz}, dt::f32, tag::nchw);
    auto conv_bias_md = memory::desc({conv_bias_tz}, dt::f32, tag::x);
    auto conv_weights_md = memory::desc({conv_weights_tz}, dt::f32, tag::oihw);
    auto conv_dst_md = memory::desc({conv_dst_tz}, dt::f32, tag::nchw);

    // init memory
    auto user_src_memory = sycl_interop::make_memory(conv_src_md, eng, sycl_interop::memory_kind::buffer);
    auto user_weights_memory = sycl_interop::make_memory(conv_weights_md, eng, sycl_interop::memory_kind::buffer);
    auto user_bias_memory = sycl_interop::make_memory(conv_bias_md, eng, sycl_interop::memory_kind::buffer);   
    auto conv_dst_memory = sycl_interop::make_memory(conv_dst_md, eng, sycl_interop::memory_kind::buffer);

    // create convolution_forward description
    auto conv_desc = convolution_forward::desc(prop_kind::forward_inference,
        algorithm::convolution_direct, conv_src_md, conv_weights_md,
        conv_bias_md, conv_dst_md, conv_strides, conv_padding,
        conv_padding);

    // create primitive description (convolution forward)
    auto conv_prim_desc = convolution_forward::primitive_desc(conv_desc, eng);    
    
    // create primitive (convolution forward)
    auto conv = convolution_forward(conv_prim_desc);

    // data buffer
    write_to_dnnl_memory(src_v.data(), user_src_memory);
    write_to_dnnl_memory(weight_v.data(), user_weights_memory);
    write_to_dnnl_memory(bias_v.data(), user_bias_memory);

    // forward propagate
    conv.execute(
        engine_stream,
        {
            {DNNL_ARG_SRC, user_src_memory},
            {DNNL_ARG_WEIGHTS, user_weights_memory},
            {DNNL_ARG_BIAS, user_bias_memory},
            {DNNL_ARG_DST, conv_dst_memory}
        }
    );
    
    engine_stream.wait();

    read_from_dnnl_memory(dst_v.data(), conv_dst_memory);
    printf("last ele: %f\n", dst_v[dst_v.size()-1]);
}

int main(int argc, char **argv) {
    engine::kind engine_kind = dnnl::engine::kind::gpu;
    convtest(engine_kind);
}
```
I run it on a cuda backend onednn, 
the result of it is:
```
input tensor: 100×3×227×227
filter: 96×3×11×11
last ele: 36.300018
```
I was expecting a float number exactly equals to 36.3, which I have testified on a simple cuda code. 
the cuda code is as in below:
```cuda
#include <iostream>
#include <cudnn.h>
#include <vector>

using namespace std;

int main()
{   
    int arg[7] = {3, 227, 227, 11, 4, 96, 0};

    int batch=100, input_c=arg[0], input_h=arg[1], input_w=arg[2];
    int filter_size=arg[3], stride=arg[4], padding=arg[6];
    int output_c=arg[5];
    int output_h = (input_h - filter_size + 2*padding) / stride + 1;
    int output_w = (input_w - filter_size + 2*padding) / stride + 1;
    
    float *ds, *dw, *dd;

    // host data
    std::vector<float> src_v(batch * input_c * input_h * input_w);
    std::vector<float> dst_v(batch * output_c * output_h * output_w);
    std::vector<float> weight_v(output_c * input_c * filter_size * filter_size);
    // std::vector<float> bias_v(output_c);

    for (size_t i = 0; i < src_v.size(); ++i)
        src_v[i] = 1.0f;
    for (size_t i = 0; i < weight_v.size(); ++i)
        weight_v[i] = 0.1f;
    cudaMalloc(&ds, src_v.size()*sizeof(float));
    cudaMalloc(&dw, weight_v.size()*sizeof(float));
    cudaMalloc(&dd, dst_v.size()*sizeof(float));

    cudaMemcpy(ds, src_v.data(), src_v.size()*sizeof(float), cudaMemcpyHostToDevice);
    cudaMemcpy(dw, weight_v.data(), weight_v.size()*sizeof(float), cudaMemcpyHostToDevice);
    cudaDeviceSynchronize();

    cudnnHandle_t handle;
    cudnnCreate(&handle);

    cudnnTensorDescriptor_t src_d, dst_d;
    cudnnFilterDescriptor_t weight_d;

    cudnnCreateTensorDescriptor(&src_d);
    cudnnCreateTensorDescriptor(&dst_d);
    cudnnCreateFilterDescriptor(&weight_d);

    cudnnSetTensor4dDescriptor(src_d, CUDNN_TENSOR_NCHW, cudnnDataType_t::CUDNN_DATA_FLOAT, batch, input_c, input_h, input_w);
    cudnnSetTensor4dDescriptor(dst_d, CUDNN_TENSOR_NCHW, CUDNN_DATA_FLOAT, batch, output_c, output_h, output_w);
    cudnnSetFilter4dDescriptor(weight_d, CUDNN_DATA_FLOAT, CUDNN_TENSOR_NCHW, output_c, input_c, filter_size, filter_size);

    cudnnConvolutionDescriptor_t conv_d;
    cudnnCreateConvolutionDescriptor(&conv_d);
    cudnnSetConvolution2dDescriptor(conv_d, padding, padding, stride, stride, 1, 1, cudnnConvolutionMode_t::CUDNN_CONVOLUTION, CUDNN_DATA_FLOAT);

    int max_count, useable_count;
    std::vector<cudnnConvolutionFwdAlgoPerf_t> perf;
    cudnnGetConvolutionForwardAlgorithmMaxCount(handle, &max_count);
    perf.resize(max_count);
    cudnnFindConvolutionForwardAlgorithm(handle, src_d, weight_d, conv_d, dst_d, max_count, &useable_count, perf.data());

    std::cout<<"input tensor: "<<batch<<", "<<input_c<<", "<<input_h<<", "<<input_w<<"\n";
    bool algo_selected=false;
    int algo_specify = 1;
    cudnnConvolutionFwdAlgoPerf_t algoP;
    for(int i=0;i<useable_count;i++){
        if((int)perf[i].algo==algo_specify){
            algo_selected=true;
            algoP = perf[i];
        }
    }
    if(!algo_selected){
        printf("algo not suppuort!\n");
    }

    size_t ws_size;
    char* ws;
    cudnnGetConvolutionForwardWorkspaceSize(handle, src_d, weight_d, conv_d, dst_d, algoP.algo, &ws_size);
    cudaMalloc(&ws, ws_size);
    cudaDeviceSynchronize(); 

    float alpha = 1.0f, beta = .0f;
    cudaDeviceSynchronize();
    cudnnConvolutionForward(handle, &alpha, src_d, ds, weight_d, dw, conv_d, algoP.algo, ws, ws_size, &beta, dst_d, dd);

    std::cout<<"algo: "<<algoP.algo<<"\n";
    
    cudaDeviceSynchronize();
    cudnnConvolutionForward(handle, &alpha, src_d, ds, weight_d, dw, conv_d, algoP.algo, ws, ws_size, &beta, dst_d, dd);
    
    cudaDeviceSynchronize();
    cudaMemcpy(dst_v.data(), dd, sizeof(float)*dst_v.size(), cudaMemcpyDeviceToHost);

    std::cout<<"last ele:"<<dst_v[dst_v.size()-1]<<"\n";
}
```

**Comment by dzarukin** [comment]
Hi @MOLOjl, thanks for your question.
It hard to expect that different hardware and software will produce same floating point results on a random set of data.
Even on the same hardware different software algorithms may produce different results. fp type is very sensible to changes in operation order, esp. for non-exact data. Additional information may be found in the Internet.

If you really want to see exactness, my recommendation would be to use **only** powers-of-two in data and check that cuda and oneDNN algorithms would be direct or gemm-like, but not fft or wino. Thanks.

**Comment by vpirogov** [comment]
@MOLOjl, do you run oneDNN on NVIDIA as well? If that's the case the difference may come from different algorithms used in computations. You can find logic used in oneDNN for cuDNN algorithm selection [here](https://github.com/oneapi-src/oneDNN/blob/d5ad699851ce02ff55353b389926caa7339fc8a9/src/gpu/nvidia/cudnn_convolution_impl.hpp#L578).

**Comment by MOLOjl** [comment]
I changed the source code and add some output commands, I measured the time cost and found the algorithm used in convolution forward, the changes in `cudnn_convolution_impl.hpp` are as follows.
```c++
// ...
if (true) {
    const float bias_alpha = 1.0f;
    const float bias_beta = 1.0f;

    std::cout << "selceted algo: " << fwd_alg_kind << "\n";
    cudaDeviceSynchronize();
    auto start = std::chrono::system_clock::now();

    CUDNN_EXECUTE_FUNC_V(cudnnConvolutionForward, handle, &alpha,
            descs[io::x], x, weights_desc, weights, conv_desc,
            fwd_alg_kind, scratchpad, scratchpad_size, &beta,
            descs[io::y], output);
    cudaDeviceSynchronize();
    auto end = std::chrono::system_clock::now();

    auto time_compute = duration_cast<std::chrono::microseconds>(end-start);
    
    std::cout << "time compute:" <<
        float(time_compute.count()) * std::chrono::microseconds::period::num/std::chrono::microseconds::period::den 
        << "\n";
}
// ...
```
the output shows cuda program and c++ program(using onednn) uses the same algorithm
 (seems `CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM`), 
and the similar time cost evidenced it.
I still can't find why, since they all use `cudnnConvolutionForward()` to do the computation, maybe there this something I missed, or just because of simple negligence, I will try more to find out why this happen.
I don't know whether you have done some similar tests and show the same problem, if your test has no precision lost, please let me know, it will prove that my code must be wrong somewhere.

**Comment by dzarukin** [comment]
Hi @MOLOjl,
I see that cuDNN provides CUDNN_LOGINFO_DBG variable to log API calls. Could you use that to verify that call are absolutely identical, please?
From what I see we don't use `cudnnSetTensor4dDescriptor` or `cudnnSetFilter4dDescriptor`. We use `cudnnSetTensor`. I don't know if this could lead to some internal modifications to the flow that cuDNN may convert internally to some different format, changing the order of operation or something like that.

I may propose to change `0.1f` value in weights to some power of 2, like `0.25f` or `0.125f` since this number has exact representation as floating-point value, while `0.1f` does not.
Thanks.

**Comment by MOLOjl** [comment]
@dzarukin  
sorry for wasting your time, I found the reason, it's just a silly mistake. 
I used two different ways to output the result (`std::cout` and `print()`), which caused this problem.
I searched on the internet and got a conclusion that seems right.

float number only has 6 significant figures that are guaranteed absolutely correct (2^23 = 8388608), so `std::cout` may calculate out 6 significant figures from the original binary number and ignore the 0 in the tail.
However, `printf()` using `%f` choose to output six numbers after the decimal point.

so caused this difference.
